File: dto-mapping-benchmark/src/main/java/com/example/dtomappingbenchmark/v2/service/UserService.java
package com.example.dtomappingbenchmark.v2.service;

import com.example.dtomappingbenchmark.v1.entity.Post;
import com.example.dtomappingbenchmark.v2.dto.UserDto;
import com.example.dtomappingbenchmark.v2.entity.User;
import com.example.dtomappingbenchmark.v2.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final PostService postService;

    public UserDto create(UserDto dto) {
        User user = toEntity(dto);
        return toDto(userRepository.save(user));
    }

    @Transactional(readOnly = true)
    public UserDto get(Long id) {
        User user = userRepository.findById(id).orElseThrow();
        return toDto(user);
    }

    private UserDto toDto(User user) {
        return UserDto.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .postList(user.getPosts().stream()
                        .map(postService::toDto)
                        .toList())
                .build();
    }

    public User toEntity(UserDto dto){
        return User.builder()
                .name(dto.getName())
                .email(dto.getEmail())
                .posts(new ArrayList<>())
                .build();
    }
}

File: object-mapping-benchmark-v1/src/main/kotlin/yu/selab/cj/objectmappingbenchmarkv1/service/UserService.kt
package yu.selab.cj.objectmappingbenchmarkv1.service

import org.springframework.stereotype.Service
import yu.selab.cj.objectmappingbenchmarkv1.command.UserDto
import yu.selab.cj.objectmappingbenchmarkv1.entity.User
import yu.selab.cj.objectmappingbenchmarkv1.entity.UserRepository

@Service
class UserService(
    private val userRepository : UserRepository
) {

    fun getUserList() = userRepository.findAll()

    fun getUser(id : Long) = userRepository.findById(id).orElseThrow {
        IllegalArgumentException("user not found")
    }

    fun createUser(dto : UserDto){
        val user = User(dto.name, dto.password)
        userRepository.save(user)
    }

    fun deleteUser(id : Long)
        = userRepository.deleteById(id)

}